File: ruby/binary/binary.rb
class Binary
  def initialize(num)
    if num.match(/[^01]/) != nil
      raise ArgumentError.new("Only 0s and 1s allowed")
    end
    @num = num
  end

  def to_decimal
    num = @num

    if num.length <= 1
      num.to_i*(2**0)
    else
      binary = []
      num = num.split("").reverse
      num.each_with_index { |num, index|
        binary << num.to_i*(2**index)
      }
      binary.reduce(:+)
    end
  end
end

module BookKeeping
    VERSION = 2
end

File: javascript/binary/binary.js
var Binary = function(bits){
  this.bits = bits;
}

Binary.prototype.toDecimal = function(){
  var base = 2;
  var binaryArr = this.bits.split("").reverse();
  var length = binaryArr.length;

  for(var i = 0; i < length; i++){
    if(binaryArr[i] === '0' || binaryArr[i] === '1'){
      binaryArr[i] *= Math.pow(base, i);
    } else {
      return 0;
    }
  }
  
  return binaryArr.reduce(function(a,b){
    return a + b;
  }, 0);
}

module.exports = Binary;